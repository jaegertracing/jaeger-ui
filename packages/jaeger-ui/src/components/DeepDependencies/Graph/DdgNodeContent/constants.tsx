// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { getUrl as getSearchUrl } from '../../../SearchTracePage/url';

// Constants for calculating circular node positioning
export const FONT_SIZE = 14;
export const FONT = `${FONT_SIZE}px Helvetica Nueue`;
export const LINE_HEIGHT = 1.5;
export const OP_PADDING_TOP = 5;
export const PROGRESS_BAR_STROKE_WIDTH = 15;
export const RADIUS = 75;
export const WORD_RX = /\W*\w+\W*/g;

// While browsers suport URLs of unlimited length, many server clients do not handle more than this max
// export const MAX_LENGTH = 2083;
// get unpreditable, unreproducable results past 7400 characters
export const MAX_LENGTH = 7000;
export const MAX_LINKED_TRACES = 35;
export const MIN_LENGTH = getSearchUrl().length;
export const PARAM_NAME_LENGTH = '&traceID='.length;
