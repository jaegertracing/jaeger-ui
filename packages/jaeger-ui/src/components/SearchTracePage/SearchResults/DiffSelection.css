/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DiffSelection.is-non-empty {
  position: sticky;
  top: 47px;
  z-index: 1;
}

.DiffSelection--message {
  background: #f5f5f5;
  z-index: 11;
  background: #e5f2f2;
  border: 1px solid #87c3c3;
  padding: 1rem;
}

.DiffSelection--selectedItems {
  border: 1px solid #ccc;
  border-bottom: none;
}

.DiffSelection--message button {
  margin-left: 7px;
}
