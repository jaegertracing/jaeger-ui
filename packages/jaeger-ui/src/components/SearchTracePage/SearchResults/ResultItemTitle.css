/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.ResultItemTitle {
  background: #ececec;
  border-bottom: 1px solid #d8d8d8;
  display: flex;
}

.ResultItemTitle--item {
  color: inherit;
  padding: 0.5rem;
  position: relative;
}

.ResultItemTitle--item:first-child {
  border-right: 1px solid #ddd;
}

.ResultItemTitle--item:hover {
  background: #e4e4e4;
  border-color: #ccc;
}

.ResultItemTitle--title {
  margin: 0;
  position: relative;
}

.ResultItemTitle--title.is-error {
  color: #c00;
}

.ResultItemTitle--idExcerpt {
  color: #888;
  font-weight: normal;
  padding-left: 0.5rem;
}

.ResultItemTitle--durationBar {
  background: #d7e7ea;
  bottom: 0;
  left: 0;
  position: absolute;
  top: 0;
}

.ResultItemTitle--item:hover > .ResultItemTitle--durationBar {
  background: #c5dde0;
}
