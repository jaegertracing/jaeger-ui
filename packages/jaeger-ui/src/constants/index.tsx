// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export const TOP_NAV_HEIGHT = 47 as 47;

export const FALLBACK_DAG_MAX_NUM_SERVICES = 100 as 100;
export const FALLBACK_TRACE_NAME = '<trace-without-root-span>' as '<trace-without-root-span>';
export const FALLBACK_STANDARD_DATETIME_FORMAT = 'LLL';

export const FETCH_DONE = 'FETCH_DONE' as 'FETCH_DONE';
export const FETCH_ERROR = 'FETCH_ERROR' as 'FETCH_ERROR';
export const FETCH_LOADING = 'FETCH_LOADING' as 'FETCH_LOADING';

export const fetchedState = {
  DONE: FETCH_DONE,
  ERROR: FETCH_ERROR,
  LOADING: FETCH_LOADING,
};
