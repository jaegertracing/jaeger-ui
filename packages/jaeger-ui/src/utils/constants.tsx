// Copyright (c) 2023 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Provides access to constants injected by the build system.
 */

/**
 * Get the current execution environment, as inferred from NODE_ENV at build time.
 */
export function getAppEnvironment() {
  return __APP_ENVIRONMENT__;
}

/**
 * Get injected version details as a JSON-formatted string.
 */
export function getVersionInfo() {
  return __REACT_APP_VSN_STATE__;
}

export function shouldDebugGoogleAnalytics() {
  return __REACT_APP_GA_DEBUG__;
}
