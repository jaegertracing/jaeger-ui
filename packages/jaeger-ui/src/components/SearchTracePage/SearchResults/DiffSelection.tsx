// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Button } from 'antd';
import { Link } from 'react-router-dom';

import ResultItemTitle from './ResultItemTitle';
import { getUrl } from '../../TraceDiff/url';
import { getUrl as getTracePageUrl } from '../../TracePage/url';
import { fetchedState } from '../../../constants';

import { FetchedTrace } from '../../../types';

import './DiffSelection.css';

type Props = {
  toggleComparison: (traceID: string, isInDiffCohort: boolean) => void;
  traces: FetchedTrace[];
};

// Exported for tests
export const CTA_MESSAGE = <h2 className="ub-m0">Compare traces by selecting result items</h2>;

export default class DiffSelection extends React.PureComponent<Props> {
  render() {
    const { toggleComparison, traces } = this.props;
    const cohort = traces.filter(ft => ft.state !== fetchedState.ERROR).map(ft => ft.id);
    const compareHref = cohort.length > 1 ? getUrl({ cohort }) : null;
    const compareBtn = (
      <Button className="ub-right" disabled={cohort.length < 2} htmlType="button" type="primary">
        Compare Traces
      </Button>
    );
    return (
      <div className={`DiffSelection ${traces.length ? 'is-non-empty' : ''} ub-mb3`}>
        {traces.length > 0 && (
          <div className="DiffSelection--selectedItems">
            {traces.map(fetchedTrace => {
              const { data, error, id, state } = fetchedTrace;
              return (
                <ResultItemTitle
                  key={id}
                  duration={data && data.duration}
                  error={error}
                  isInDiffCohort
                  linkTo={getTracePageUrl(id)}
                  state={state}
                  targetBlank
                  toggleComparison={toggleComparison}
                  traceID={id}
                  traceName={data && data.traceName}
                />
              );
            })}
          </div>
        )}
        <div className="DiffSelection--message">
          {traces.length > 0 ? (
            <React.Fragment>
              {compareHref ? <Link to={compareHref}>{compareBtn}</Link> : compareBtn}
              <h2 className="ub-m0">{cohort.length} Selected for comparison</h2>
            </React.Fragment>
          ) : (
            CTA_MESSAGE
          )}
        </div>
      </div>
    );
  }
}
