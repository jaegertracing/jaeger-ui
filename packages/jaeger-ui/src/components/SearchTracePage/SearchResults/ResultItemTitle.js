// @flow

// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Checkbox } from 'antd';
import { Link } from 'react-router-dom';

import TraceName from '../../common/TraceName';
import { fetchedState } from '../../../constants';
import { formatDuration } from '../../../utils/date';

import type { FetchedState } from '../../../types';
import type { ApiError } from '../../../types/api-error';

import './ResultItemTitle.css';

type Props = {
  duration: number,
  durationPercent: number,
  error?: ApiError,
  isInDiffCohort: boolean,
  linkTo: ?string,
  state: ?FetchedState,
  toggleComparison: (string, boolean) => void,
  traceID: string,
  traceName: string,
  disableComparision?: boolean,
};

export default class ResultItemTitle extends React.PureComponent<Props> {
  props: Props;

  static defaultProps = {
    durationPercent: 0,
    error: undefined,
    state: fetchedState.DONE,
    linkTo: null,
  };

  toggleComparison = () => {
    const { isInDiffCohort, toggleComparison, traceID } = this.props;
    toggleComparison(traceID, isInDiffCohort);
  };

  render() {
    const {
      duration,
      durationPercent,
      error,
      isInDiffCohort,
      linkTo,
      state,
      traceID,
      traceName,
      disableComparision,
    } = this.props;
    // Use a div when the ResultItemTitle doesn't link to anything
    let WrapperComponent = 'div';
    const wrapperProps: { [string]: string } = { className: 'ResultItemTitle--item ub-flex-auto' };
    if (linkTo) {
      WrapperComponent = Link;
      wrapperProps.to = linkTo;
    }
    const isErred = state === fetchedState.ERROR;
    return (
      <div className="ResultItemTitle">
        {!disableComparision && (
          <Checkbox
            className="ResultItemTitle--item ub-flex-none"
            checked={!isErred && isInDiffCohort}
            disabled={isErred}
            onChange={this.toggleComparison}
          />
        )}
        <WrapperComponent {...wrapperProps}>
          <span className="ResultItemTitle--durationBar" style={{ width: `${durationPercent}%` }} />
          {duration != null && <span className="ub-right ub-relative">{formatDuration(duration)}</span>}
          <h3 className="ResultItemTitle--title">
            <TraceName error={error} state={state} traceName={traceName} />
            <small className="ResultItemTitle--idExcerpt">{traceID.slice(0, 7)}</small>
          </h3>
        </WrapperComponent>
      </div>
    );
  }
}
