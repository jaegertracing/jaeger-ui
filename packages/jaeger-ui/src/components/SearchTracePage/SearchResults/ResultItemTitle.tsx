// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Checkbox } from 'antd';
import { LocationDescriptor } from 'history';
import { Link } from 'react-router-dom';

import TraceName from '../../common/TraceName';
import { fetchedState } from '../../../constants';
import { formatDuration } from '../../../utils/date';

import { FetchedState, TNil } from '../../../types';
import { ApiError } from '../../../types/api-error';

import './ResultItemTitle.css';
import { getTargetEmptyOrBlank } from '../../../utils/config/get-target';

type Props = {
  duration?: number;
  durationPercent?: number;
  error?: ApiError;
  isInDiffCohort: boolean;
  linkTo: LocationDescriptor | TNil;
  state?: FetchedState | TNil;
  targetBlank?: boolean;
  toggleComparison: (traceID: string, isInDiffCohort: boolean) => void;
  traceID: string;
  traceName?: string;
  disableComparision?: boolean;
};

const DEFAULT_DURATION_PERCENT = 0;

const stopCheckboxPropagation = (evt: React.MouseEvent) => evt.stopPropagation();

export default class ResultItemTitle extends React.PureComponent<Props> {
  static defaultProps: Partial<Props> = {
    disableComparision: false,
    durationPercent: DEFAULT_DURATION_PERCENT,
    error: undefined,
    state: undefined,
    targetBlank: false,
  };

  toggleComparison = () => {
    const { isInDiffCohort, toggleComparison, traceID } = this.props;
    toggleComparison(traceID, isInDiffCohort);
  };

  render() {
    const {
      disableComparision,
      duration,
      durationPercent,
      error,
      isInDiffCohort,
      linkTo,
      state,
      targetBlank,
      traceID,
      traceName,
    } = this.props;
    // Use a div when the ResultItemTitle doesn't link to anything
    let WrapperComponent: string | typeof Link = 'div';
    const wrapperProps: Record<string, string | LocationDescriptor> = {
      className: 'ResultItemTitle--item ub-flex-auto',
    };
    if (linkTo) {
      wrapperProps.to = linkTo;
      WrapperComponent = Link;
      if (targetBlank) {
        wrapperProps.target = getTargetEmptyOrBlank();
        wrapperProps.rel = 'noopener noreferrer';
      }
    }
    const isErred = state === fetchedState.ERROR;
    // Separate propagation management and toggle manegement due to ant-design#16400
    const checkboxProps = {
      className: 'ResultItemTitle--item ub-flex-none',
      checked: !isErred && isInDiffCohort,
      disabled: isErred,
      onChange: this.toggleComparison,
      onClick: stopCheckboxPropagation,
    };

    return (
      <div className="ResultItemTitle">
        {!disableComparision && <Checkbox {...checkboxProps} />}
        {/* TODO: Shouldn't need cast */}
        <WrapperComponent {...(wrapperProps as { [key: string]: string; to: string })}>
          <span
            className="ResultItemTitle--durationBar"
            style={{ width: `${durationPercent || DEFAULT_DURATION_PERCENT}%` }}
          />
          {duration != null && <span className="ub-right ub-relative">{formatDuration(duration)}</span>}
          <h3 className="ResultItemTitle--title">
            <TraceName error={error} state={state} traceName={traceName} />
            <small className="ResultItemTitle--idExcerpt">{traceID.slice(0, 7)}</small>
          </h3>
        </WrapperComponent>
      </div>
    );
  }
}
