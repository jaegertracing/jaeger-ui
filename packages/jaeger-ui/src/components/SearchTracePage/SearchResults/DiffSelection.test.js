// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';

import DiffSelection, { CTA_MESSAGE } from './DiffSelection.js';
import ResultItemTitle from './ResultItemTitle';
import { fetchedState } from '../../../constants';

describe('DiffSelection', () => {
  const toggleComparison = () => {};
  const traces = [
    {
      id: 'trace-id-0',
      data: {
        duration: 0,
        traceName: 'trace-name-0',
      },
      error: new Error('error-0'),
      state: fetchedState.DONE,
    },
    {
      id: 'trace-id-1',
      // deliberately missing data to test default
      error: new Error('error-1'),
      state: fetchedState.DONE,
    },
    {
      id: 'trace-id-2',
      // deliberately missing data to test default
      error: new Error('error-2'),
      state: fetchedState.ERROR,
    },
  ];

  it('renders a trace as expected', () => {
    const wrapper = shallow(
      <DiffSelection traces={traces.slice(0, 1)} toggleComparison={toggleComparison} />
    );

    expect(wrapper.find(ResultItemTitle).length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders multiple traces as expected', () => {
    const wrapper = shallow(<DiffSelection traces={traces} toggleComparison={toggleComparison} />);

    expect(wrapper.find(ResultItemTitle).length).toBe(traces.length);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders CTA_MESSAGE when given empty traces array', () => {
    const wrapper = shallow(<DiffSelection traces={[]} toggleComparison={toggleComparison} />);
    expect(wrapper.contains(CTA_MESSAGE)).toBe(true);
  });
});
