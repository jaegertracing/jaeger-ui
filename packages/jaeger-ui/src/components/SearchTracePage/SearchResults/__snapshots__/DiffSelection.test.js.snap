// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffSelection renders a trace as expected 1`] = `
<div
  className="DiffSelection is-non-empty ub-mb3"
>
  <div
    className="DiffSelection--selectedItems"
  >
    <ResultItemTitle
      disableComparision={false}
      duration={0}
      durationPercent={0}
      error={[Error: error-0]}
      isInDiffCohort={true}
      key="trace-id-0"
      linkTo="/trace/trace-id-0"
      state="FETCH_DONE"
      targetBlank={true}
      toggleComparison={[Function]}
      traceID="trace-id-0"
      traceName="trace-name-0"
    />
  </div>
  <div
    className="DiffSelection--message"
  >
    <Button
      block={false}
      className="ub-right"
      disabled={true}
      ghost={false}
      htmlType="button"
      loading={false}
      prefixCls="ant-btn"
      type="primary"
    >
      Compare Traces
    </Button>
    <h2
      className="ub-m0"
    >
      1
       Selected for comparison
    </h2>
  </div>
</div>
`;

exports[`DiffSelection renders multiple traces as expected 1`] = `
<div
  className="DiffSelection is-non-empty ub-mb3"
>
  <div
    className="DiffSelection--selectedItems"
  >
    <ResultItemTitle
      disableComparision={false}
      duration={0}
      durationPercent={0}
      error={[Error: error-0]}
      isInDiffCohort={true}
      key="trace-id-0"
      linkTo="/trace/trace-id-0"
      state="FETCH_DONE"
      targetBlank={true}
      toggleComparison={[Function]}
      traceID="trace-id-0"
      traceName="trace-name-0"
    />
    <ResultItemTitle
      disableComparision={false}
      durationPercent={0}
      error={[Error: error-1]}
      isInDiffCohort={true}
      key="trace-id-1"
      linkTo="/trace/trace-id-1"
      state="FETCH_DONE"
      targetBlank={true}
      toggleComparison={[Function]}
      traceID="trace-id-1"
    />
    <ResultItemTitle
      disableComparision={false}
      durationPercent={0}
      error={[Error: error-2]}
      isInDiffCohort={true}
      key="trace-id-2"
      linkTo="/trace/trace-id-2"
      state="FETCH_ERROR"
      targetBlank={true}
      toggleComparison={[Function]}
      traceID="trace-id-2"
    />
  </div>
  <div
    className="DiffSelection--message"
  >
    <Link
      replace={false}
      to="/trace/trace-id-0...trace-id-1?cohort=trace-id-0&cohort=trace-id-1"
    >
      <Button
        block={false}
        className="ub-right"
        disabled={false}
        ghost={false}
        htmlType="button"
        loading={false}
        prefixCls="ant-btn"
        type="primary"
      >
        Compare Traces
      </Button>
    </Link>
    <h2
      className="ub-m0"
    >
      2
       Selected for comparison
    </h2>
  </div>
</div>
`;
